### ISSUE TITLE
Proposal: Experimental Differentiable Backend Interface (PyTorch/Autograd Bridge)

### ISSUE THREAD

**Issue by hillsberg3008** [brian-team/brian2#1738]
Hi Brian Team,

I have been exploring `brian2` code generation backends . Currently, advanced parameter estimation relies heavily on gradient-free optimizers (like Nevergrad) or external tools (SBI), which can be sample-inefficient for high-dimensional parameter spaces compared to direct gradient-based optimization.

The Proposal
I would like to propose (and implement) a lightweight "Differentiable Shim" for Brian2 as a GSoC 2026 project.
Instead of rewriting the core in JAX/PyTorch immediately (which is a massive undertaking), we can wrap the existing `Network.run()` in a custom `torch.autograd.Function`.

Technical Approach:
1.  Forward Pass: Executes the standard Brian2 code (leveraging the existing efficient C++ generation).
2.  Backward Pass: Uses a user-definable "Surrogate Gradient" (e.g., Fast Sigmoid or SuperSpike) to approximate gradients at spike times. This allows `loss.backward()` to flow from output traces/spike trains back to the model parameters.
3.  Integration: This would enable users to optimize Brian2 neuron parameters using standard PyTorch optimizers (Adam, SGD) directly, without leaving the Brian2 ecosystem.

Proof of My Concept:
I have verified this approach locally on my machine. I successfully wrapped a standard LIF simulation in `torch.autograd`, allowing a PyTorch optimizer to recover the membrane time constant `tau` by minimizing trace error via backpropagation.

I believe this infrastructure is the missing link to modernize parameter fitting in Brian2 and enable Scientific ML workflows.

My Question:
Is this "Wrapper" approach aligned with the core team's vision for differentiability, or would you prefer a deeper integration (e.g., a dedicated `torch` device backend)? I would love to contribute this feature.

**Comment by mstimberg** [comment]
Hi @hillsberg3008. As I mentioned in the brian2modelfitting PRs, I've been busy pushing out Brian 2.10 lately, apologies for the late reply. This is a non-trivial question that merits a thoughtful response, I hope I get around to providing one before the end of the week (if not, please ping me here…).

**Comment by hillsberg3008** [comment]
Hi @mstimberg . I have been waiting for your response for a long time now. I know you must be having a busy schedule . I will appreciate if you reconsider and review my approach for once. 
Thanks for your time.

**Comment by mstimberg** [comment]
Dear @hillsberg3008. Thank you for reminding me, and apologies for the late reply, the end of the year was indeed a bit busy for me...
In general, I'd be very happy with any kind of "lightweight" solution, in particular something that could be implemented outside of Brian itself, say a `autograd_run` function that wraps `Network.run`. But I have to admit I do not quite understand how the proposed approach would work here – this might also be due to my very basic experience with pytorch, so please correct any misunderstanding on my part.
From what I understand, `pytorch.autograd` works by attaching gradient information to each tensor, and updates it for each operation on it (or rather first records the computational graph, and then does these updates when you ask for a backward pass). You can then use this information to optimize with respect to some parameter. I don't see that we can directly apply this with a "wrapper" to Brian since we are not using pytorch tensors in the first place. For the numpy backend, this would be doable in principle, though, and we could also implement our custom functions in a way that could hook into pytorch's autograd mechanism. But all this wouldn't be possible for the C++/CUDA code generation targets, where the generated code operates on straightforward double arrays. Finally, Brian's generated code updates variables in a way that is not compatible with pytorch's autograd  (I think). E.g., let's say we use the `exact` method to integrate `dv/dt = -v/tau`, then the code we generate will look like this:
```python
v_new = v*exp(-tau/dt)
v[:] = v_new
```
With all this, I could potentially see a way to make Brian's `numpy` backend compatible with `pytorch.autograd` (which basically means replacing it with a `pytorch` backend), at least for a subset of Brian's functionality, but it would be a lot of work. I don't see how to do this for the C++ code generation, except for basically rewriting a lot if from scratch.

When you say,

> I have verified this approach locally on my machine. I successfully wrapped a standard LIF simulation in `torch.autograd`, allowing a PyTorch optimizer to recover the membrane time constant `tau` by minimizing trace error via backpropagation.

do you mean that you wrapped a LIF simulation that you wrote by hand? Or do you mean that you wrapped a Brian simulation? If it is the latter, that would most likely mean that my thinking detailed above is incorrect, and I'd be very interested in seeing the code!
